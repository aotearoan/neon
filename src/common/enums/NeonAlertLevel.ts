/**
 * Alert levels are the severity levels to be used with <a href="/feedback/alert#examples">NeonAlert & NeonToast</a>
 * components to inform the user of the severity of a message.
 */
export enum NeonAlertLevel {
  /** Information level, use for standard informational messages. */
  Info = 'info',
  /** Success level, use for indicating successful actions. */
  Success = 'success',
  /** Warning level, use for indicating a problem that the user <strong>can</strong> recover from. */
  Warn = 'warn',
  /** Error level, use for indicating a problem that the user <strong>cannot</strong> recover from. */
  Error = 'error',
}
