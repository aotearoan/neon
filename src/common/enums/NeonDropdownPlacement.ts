/**
 * Describes placement of dropdown related components like:
 * <ul class="no-style">
 *  <li><a href="/user-input/select">NeonSelect</a></li>
 *  <li><a href="/navigation/dropdown-menu">NeonDropdownMenu</a></li>
 *  <li><a href="/presentation/dropdown">NeonDropdown</a></li>
 * </ul>
 *
 * <p>The placement has a <strong>primary placement</strong> (the first direction) & a <strong>secondary placement</strong>
 * (the second direction). The primary placement is the location of the popup relative to the triggering element (e.g.
 * the select button). With the secondary placement defining the alignment relative to the trigger. E.g. BottomLeft will
 * open the dropdown below the trigger element with the left-hand edges aligned.</p>
 *
 * <p>The placement utilities are smart & dynamic. They listen to resize events & aim for best effort based on where the
 * dropdown is opened. E.g. in the case above with BottomLeft if there is not enough visible room to align the component
 * to the left then it will try to align to the right & likewise if there is not enough space below the trigger it will
 * align to the top (& reverse the option order in the case of a select).</p>
 */
export enum NeonDropdownPlacement {
  /** Placement above trigger element, aligned to the left */
  TopLeft = 'top-left',
  /** Placement above trigger element, aligned to the right */
  TopRight = 'top-right',
  /** Placement below trigger element, aligned to the left */
  BottomLeft = 'bottom-left',
  /** Placement below trigger element, aligned to the right */
  BottomRight = 'bottom-right',
  /** Placement to the left of the trigger element, aligned to the top */
  LeftTop = 'left-top',
  /** Placement to the left of the trigger element, aligned to the bottom */
  LeftBottom = 'left-bottom',
  /** Placement to the right of the trigger element, aligned to the top */
  RightTop = 'right-top',
  /** Placement to the right of the trigger element, aligned to the bottom */
  RightBottom = 'right-bottom',
}
