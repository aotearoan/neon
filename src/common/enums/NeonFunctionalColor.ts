/**
 * The core color palettes available for styling components in Neon. See the design section
 * <a href="/design/color">Color</a> page for more details on color palettes.
 */
export enum NeonFunctionalColor {
  /** Low contrast is a more subtle neutral palette which can be used for secondary actions & supporting text. */
  LowContrast = 'low-contrast',
  /** Neutral is the base greyscale palette. */
  Neutral = 'neutral',
  /**
   * High contrast is a higher contrast palette than can be used for increased accessibility or for a stronger black &
   * white application.
   * */
  HighContrast = 'high-contrast',
  /** Brand is the main brand color palette, this is useful for branded CTAs, text & controls. */
  Brand = 'brand',
  /** Primary is the main application color palette. For most applications this will be used by default. */
  Primary = 'primary',
  /** Info can be used to convey informational context, e.g. notes & alerts */
  Info = 'info',
  /** Success can be used to convey successful actions. */
  Success = 'success',
  /** Warn can be used to convey issues/actions that are less severe but may still impact the user. */
  Warn = 'warn',
  /** Error can be used to convey issues/actions that are severe & that the user may not be able to recover from. */
  Error = 'error',
}
