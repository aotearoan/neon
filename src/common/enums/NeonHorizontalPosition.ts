/**
 * Describes the horizontal positioning of certain elements in the context of a component. An example would be an icon
 * in a button or label as well as the label for a checkbox or switch.
 */
export enum NeonHorizontalPosition {
  /** Left horizontal positioning. */
  Left = 'left',
  /** Right horizontal positioning. */
  Right = 'right',
}
