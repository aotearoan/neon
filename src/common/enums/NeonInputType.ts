/**
 * Corresponds to the <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types">HTML input
 * types</a> for use with <a href="/user-input/input">NeonInput</a>.
 *
 * NOTE: prefer the specific components for Number, Password, Date, File, Search fields.
 */
export enum NeonInputType {
  /** Date input for selecting dates from a calendar. Use <a href="/user-input/date-picker">NeonDatePicker</a>. */
  Date = 'date',
  /** Date time input for selecting dates including a specific time. */
  DateTime = 'datetime-local',
  /** Email input for entering email addresses. */
  Email = 'email',
  /**
   * File input for opening a file dialog. Use <a href="/user-input/file">NeonDatePicker</a> or
   * <a href="/user-input/drop-zone">NeonDropZone</a>.
   */
  File = 'file',
  /** Number input for entering numeric values. Use <a href="/user-input/number">NeonNumber</a> */
  Number = 'number',
  /** Password input for entering passwords. Use <a href="/user-input/password">NeonPassword</a> */
  Password = 'password',
  /**
   * Range input for a slider control. Use <a href="/user-input/slider">NeonSlider</a> or
   * <a href="/user-input/range-slider">NeonRangeSlider</a>
   */
  Range = 'range',
  /** Search input for searching. Use <a href="/user-input/search">NeonSearch</a> */
  Search = 'search',
  /** Telephone number input. */
  Tel = 'tel',
  /** Text input for single or multiline text input. */
  Text = 'text',
  /** Time input. */
  Time = 'time',
  /** URL input for entering web addresses. */
  Url = 'url',
}
