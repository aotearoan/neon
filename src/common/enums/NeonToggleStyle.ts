/**
 * Describes the available <a href="/user-input/toggle">NeonToggle</a> styles.
 */
export enum NeonToggleStyle {
  /** Toggle style which is a horizontal group of connected buttons with one active member. */
  Toggle = 'toggle',
  /** Traditional HTML radio button input style. */
  RadioButtons = 'radio-buttons',
}
