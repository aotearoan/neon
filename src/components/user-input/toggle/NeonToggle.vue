<template>
  <div
    :class="[
      `neon-toggle--${toggleStyle}`,
      `neon-toggle--${disabled ? 'low-contrast' : color}`,
      `neon-toggle--${orientation}`,
      `neon-toggle--${size}`,
      { 'neon-toggle--disabled': disabled },
    ]"
    class="neon-toggle"
    role="radiogroup"
  >
    <label
      v-for="option in model"
      :key="option.key"
      :aria-checked="option.key === modelValue"
      :aria-disabled="disabled || option.disabled"
      :class="{
        'neon-toggle__label--disabled': disabled || option.disabled,
        'neon-toggle__label--checked': option.key === modelValue,
        'neon-toggle__label--with-icon': option.icon,
      }"
      :tabindex="!disabled && !option.disabled ? 0 : undefined"
      class="neon-toggle__label no-style"
      role="radio"
      @keydown.enter="selectOption(option)"
      @keydown.space="selectOption(option)"
      @keypress.space.prevent=""
    >
      <input
        :checked="option.key === modelValue"
        :disabled="disabled || option.disabled"
        :name="name"
        :tabindex="-1"
        :value="option.key"
        class="neon-toggle__input"
        type="radio"
        v-bind="sanitizedAttributes"
        @click="onInput(option.key)"
      />
      <div v-if="toggleStyle === 'radio-buttons'" class="neon-toggle__radio-button">
        <div v-if="option.key === modelValue" class="neon-toggle__radio-button-indicator"></div>
      </div>
      <neon-icon v-if="option.icon" :disabled="disabled || option.disabled" :name="option.icon" />
      <span v-if="option.label">{{ option.label }}</span>
    </label>
  </div>
</template>

<script lang="ts" src="./NeonToggle.ts" />
