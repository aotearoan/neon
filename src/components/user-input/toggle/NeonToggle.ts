import { computed, defineComponent } from 'vue';
import type { NeonToggleModel } from '@/common/models/NeonToggleModel';
import { NeonSize } from '@/common/enums/NeonSize';
import { NeonFunctionalColor } from '@/common/enums/NeonFunctionalColor';
import { NeonToggleStyle } from '@/common/enums/NeonToggleStyle';
import { NeonOrientation } from '@/common/enums/NeonOrientation';
import NeonIcon from '@/components/presentation/icon/NeonIcon.vue';

/**
 * <p>A toggle component for selecting one value from a range of options. This is equivalent to a radio button group. It can be styled as a <em>Toggle</em> or as <em>Radio buttons</em>.
 */
export default defineComponent({
  name: 'NeonToggle',
  components: {
    NeonIcon,
  },
  props: {
    /**
     * The name of the radio button group.
     */
    name: { type: String, required: true },
    /**
     * The key of the selected option.
     */
    modelValue: { type: String, required: true },
    /**
     * The list of options to present to the user.
     */
    model: { type: Array as () => Array<NeonToggleModel>, required: true },
    /**
     * The style of toggle to display to the user.
     */
    toggleStyle: { type: String as () => NeonToggleStyle, default: () => NeonToggleStyle.Toggle },
    /**
     * The size of the toggle.
     */
    size: { type: String as () => NeonSize, default: () => NeonSize.Medium },
    /**
     * The orientation of the toggle if the style is a radio button group.
     */
    orientation: { type: String as () => NeonOrientation, default: () => NeonOrientation.Vertical },
    /**
     * The color of the toggle.
     */
    color: { type: String as () => NeonFunctionalColor, default: () => NeonFunctionalColor.Neutral },
    /**
     * Whether the toggle is disabled.
     */
    disabled: { type: Boolean, default: false },
  },
  emits: [
    /**
     * Emitted when the selected value changes.
     * @type {boolean} The key of the selected model item.
     */
    'update:modelValue',
  ],
  setup(props, { emit, attrs }) {
    const emitInput = (key: string) => {
      emit('update:modelValue', key);
    };

    const onInput = (key: string) => {
      if (!props.disabled) {
        emitInput(key);
      }
    };

    const sanitizedAttributes = computed(() => {
      const attributes = Object.entries(attrs).filter(([key, _value]) => key !== 'onInput' && key !== 'onClick');
      return { ...attributes };
    });

    const selectOption = (option: NeonToggleModel) => {
      if (!props.disabled && !option.disabled) {
        emitInput(option.key);
      }
    };

    return {
      selectOption,
      onInput,
      sanitizedAttributes,
    };
  },
});
