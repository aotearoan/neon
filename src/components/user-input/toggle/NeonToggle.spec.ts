import type { RenderResult } from '@testing-library/vue';
import { fireEvent, render } from '@testing-library/vue';
import NeonToggle from './NeonToggle.vue';
import { NeonToggleStyle } from '@/common/enums/NeonToggleStyle';
import { NeonSize } from '@/common/enums/NeonSize';
import { NeonFunctionalColor } from '@/common/enums/NeonFunctionalColor';
import { NeonOrientation } from '@/common/enums/NeonOrientation';

describe('NeonToggle', () => {
  const name = 'xd';

  const model = [
    {
      label: 'Option 1',
      key: 'option-1',
    },
    {
      label: 'Option 2',
      key: 'option-2',
      disabled: true,
    },
    {
      label: 'Option 3',
      key: 'option-3',
    },
  ];

  const iconModel = [
    {
      icon: 'align-left',
      key: 'left',
    },
    {
      icon: 'align-right',
      key: 'right',
    },
  ];

  const modelValue = model[0].key;
  const iconValue = iconModel[0].key;
  const props = { name, model, modelValue };

  let harness: RenderResult;

  beforeEach(() => {
    harness = render(NeonToggle, { props });
  });

  it('renders default toggle style', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatch('neon-toggle--toggle');
  });

  it('renders radio button style', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, toggleStyle: NeonToggleStyle.RadioButtons });
    // then
    expect(html()).toMatch('neon-toggle--radio-buttons');
  });

  it('renders default size', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatch('neon-toggle--m');
  });

  it('renders size', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, size: NeonSize.Large });
    // then
    expect(html()).toMatch('neon-toggle--l');
  });

  it('renders default color', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatch('neon-toggle--neutral');
  });

  it('renders color', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, color: NeonFunctionalColor.Primary });
    // then
    expect(html()).toMatch('neon-toggle--primary');
  });

  it('renders default disabled', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).not.toMatch('neon-toggle--disabled');
  });

  it('renders disabled', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, disabled: true });
    // then
    expect(html()).toMatch('neon-toggle--disabled');
  });

  it('renders icon', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, model: iconModel, modelValue: iconValue });
    // then
    expect(html()).toMatch('neon-toggle__label--with-icon');
  });

  it('renders default orientation', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatch('neon-toggle--vertical');
  });

  it('renders orientation', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, orientation: NeonOrientation.Horizontal });
    // then
    expect(html()).toMatch('neon-toggle--horizontal');
  });

  it('renders name', () => {
    // given
    const { getAllByTestId } = harness;
    // when / then
    const inputs = getAllByTestId('toggle-input');
    expect(inputs.length).toEqual(3);
    inputs.forEach((input) => expect(input.getAttribute('name')).toEqual(name));
  });

  it('renders disabled option', () => {
    // given
    const { getAllByTestId } = harness;
    // when / then
    const option = getAllByTestId('toggle-input')[1];
    expect(option.getAttribute('disabled')).toEqual('true');
    expect(option.parentElement?.getAttribute('aria-disabled')).toEqual('true');
  });

  it('renders checked option', () => {
    // given
    const { getAllByTestId } = harness;
    // when / then
    const option = getAllByTestId('toggle-input')[0];
    expect(option.getAttribute('checked')).toEqual('true');
    expect(option.parentElement?.getAttribute('aria-checked')).toEqual('true');
    expect(option.parentElement?.textContent).toEqual(model[0].label);
    expect(option.getAttribute('value')).toEqual(model[0].key);
  });

  it('renders indicator for radio buttons', async () => {
    // given
    const { html, rerender } = harness;
    // when
    await rerender({ ...props, toggleStyle: NeonToggleStyle.RadioButtons });
    // then
    expect(html()).toMatch('neon-toggle__radio-button-indicator');
  });

  it('emits selection when clicked', async () => {
    // given
    const { emitted, getAllByTestId } = harness;
    const option = getAllByTestId('toggle-input')[2];
    // when
    await fireEvent.click(option);
    // then
    expect(emitted()['update:modelValue']).toEqual([[model[2].key]]);
  });

  // it('emits selection when clicked', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   const label = wrapper.find(
  //     '.neon-toggle__label:not(.neon-toggle__label--checked):not(.neon-toggle__label--disabled)',
  //   );
  //   label.trigger('click');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue'][0]).toEqual([model[2].key]);
  // });
  //
  // it('emits nothing when clicking disabled option', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   const label = wrapper.find('.neon-toggle__label--disabled');
  //   label.trigger('click');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
  //
  // it('emits nothing when toggle disabled', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue, disabled: true },
  //   });
  //   // when
  //   const label = wrapper.findAll('.neon-toggle__label').at(0);
  //   label.trigger('click');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
  //
  // it('emits nothing when clicking checked option', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   const label = wrapper.find('.neon-toggle__label--checked');
  //   label.trigger('click');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
  //
  // it('toggles option when space pressed', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   wrapper.findAll('.neon-toggle__label').at(0).trigger('keydown.space');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue'][0]).toEqual([model[0].key]);
  // });
  //
  // it('does not toggle option when space pressed and option disabled', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   wrapper.findAll('.neon-toggle__label').at(1).trigger('keydown.space');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
  //
  // it('does not toggle option when space pressed and toggle disabled', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue, disabled: true },
  //   });
  //   // when
  //   wrapper.findAll('.neon-toggle__label').at(0).trigger('keydown.space');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
  //
  // it('toggles option when enter pressed', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   wrapper.findAll('.neon-toggle__label').at(0).trigger('keydown.enter');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue'][0]).toEqual([model[0].key]);
  // });
  //
  // it('does not toggle option when enter pressed and option disabled', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue },
  //   });
  //   // when
  //   wrapper.findAll('.neon-toggle__label').at(1).trigger('keydown.enter');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
  //
  // it('does not toggle option when enter pressed and toggle disabled', () => {
  //   // given
  //   const wrapper = shallowMount(NeonToggle, {
  //     propsData: { name, model, modelValue, disabled: true },
  //   });
  //   // when
  //   wrapper.findAll('.neon-toggle__label').at(0).trigger('keydown.enter');
  //   // then
  //   expect(wrapper.emitted()['update:modelValue']).toBeUndefined();
  // });
});
