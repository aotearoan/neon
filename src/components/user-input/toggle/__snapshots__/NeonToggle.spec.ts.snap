// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonToggle renders radio buttons with label override 1`] = `
<div class="neon-toggle--radio-buttons neon-toggle--neutral neon-toggle--vertical neon-toggle--m neon-toggle" role="radiogroup"><label aria-checked="true" class="neon-toggle__label--checked neon-toggle__label--with-slot-override neon-toggle__label--with-text neon-toggle__label no-style" tabindex="0" role="radio"><input name="xd" tabindex="-1" class="neon-toggle__input" type="radio" value="option-1">
    <div class="neon-toggle__radio-button">
      <div class="neon-toggle__radio-button-indicator"></div>
    </div><!-- @slot This slot is for overriding the option rendering, it is passed two arguments, <em>option</em> - the option model & <em>index</em> --><span>0 = description 0</span>
  </label><label aria-checked="false" aria-disabled="true" class="neon-toggle__label--disabled neon-toggle__label--with-slot-override neon-toggle__label--with-text neon-toggle__label no-style" role="radio"><input disabled="" name="xd" tabindex="-1" class="neon-toggle__input" type="radio" value="option-2">
    <div class="neon-toggle__radio-button">
      <!--v-if-->
    </div><!-- @slot This slot is for overriding the option rendering, it is passed two arguments, <em>option</em> - the option model & <em>index</em> --><span>1 = description 1</span>
  </label><label aria-checked="false" class="neon-toggle__label--with-slot-override neon-toggle__label--with-text neon-toggle__label no-style" tabindex="0" role="radio"><input name="xd" tabindex="-1" class="neon-toggle__input" type="radio" value="option-3">
    <div class="neon-toggle__radio-button">
      <!--v-if-->
    </div><!-- @slot This slot is for overriding the option rendering, it is passed two arguments, <em>option</em> - the option model & <em>index</em> --><span>2 = description 2</span>
  </label></div>
`;
